eppo_core/src/configuration_store.rs

```diff
@@ -0,0 +1,69 @@
+//! A thread-safe in-memory storage for currently active configuration. [`ConfigurationStore`]
+//! provides a concurrent access for readers (e.g., flag evaluation) and writers (e.g., periodic
+//! configuration fetcher).
+use std::sync::RwLock;
+
+use crate::Configuration;
+
+/// `ConfigurationStore` provides a thread-safe (`Sync`) storage for Eppo configuration that allows
+/// concurrent access for readers and writers.
+///
+/// `Configuration` itself is always immutable and can only be replaced fully.
+#[derive(Default)]
+pub struct ConfigurationStore {
+    configuration: RwLock<Configuration>,
+}
+
+impl ConfigurationStore {
+    pub fn new() -> Self {
+        ConfigurationStore::default()
+    }
+
+    pub fn get_configuration(&self) -> Configuration {
+        // self.configuration.read() should always return Ok(). Err() is possible only if the lock
+        // is poisoned (writer panicked while holding the lock), which should never happen.
+        let configuration = self
+            .configuration
+            .read()
+            .expect("thread holding configuration lock should not panic");
+
+        configuration.clone()
+    }
+
+    /// Set new configuration.
+    pub fn set_configuration(&self, config: Configuration) {
+        let mut configuration_slot = self
+            .configuration
+            .write()
+            .expect("thread holding configuration lock should not panic");
+
+        *configuration_slot = config;
+    }
+}
+
+#[cfg(test)]
```

Comment:
woah does rust co-locate product code and test code in the same file?

Reply:
Yes for unit-tests. Having it in the same file allows accessing private/non-exported functions in tests. `#[cfg(test)]` makes that test sub-module excluded during normal build